url969d1ddca4_compmstr_java_scheme_tgz-pJ8-com_undi_javascheme_JavaSchemeJ8/src/com/undi/util/Reflector.java

```diff
@@ -0,0 +1,371 @@
+      return argType == Float.class
+      || argType == double.class;
+    }else if(paramType == double.class){
+      return argType == Double.class
+      || argType == float.class;// || argType == DoubleNum.class;
+    }else if(paramType == long.class){
+      return argType == Long.class
+      || argType == int.class
+      || argType == short.class
+      || argType == byte.class;// || argType == BigNum.class;
+    }else if(paramType == char.class){
+      return argType == Character.class;
+    }else if(paramType == short.class){
+      return argType == Short.class;
+    }else if(paramType == byte.class){
+      return argType == Byte.class;
+    }else if(paramType == boolean.class){
+      return argType == Boolean.class;
+    }
+    return false;
+  }
+
+  static boolean isCongruent(Class<?>[] params, Object[] args){
+    boolean ret = false;
+    if(args == null)
+      return params.length == 0;
+    if(params.length == args.length){
+      ret = true;
+      for(int i = 0; ret && i < params.length; i++){
+        Object arg = args[i];
+        Class<?> argType = (arg == null) ? null : arg.getClass();
+        Class<?> paramType = params[i];
+        ret = paramArgTypeMatch(paramType, argType);
+        }
+      }
+    return ret;
+  }
+
+  
+  public static Object invokeMatchingMethod(String methodName, List<Method> methods, Object target, Object[] args){
+    Method m = null;
+    Object[] boxedArgs = null;
+    if(methods.isEmpty()){
+      throw new IllegalArgumentException("Method: '" + methodName + "' does not exist in: " + target);
+    }else if(methods.size() == 1){
+      m = (Method) methods.get(0);
+      boxedArgs = boxArgs(m.getParameterTypes(), args);
+    }else{ //overloaded with same arity
+      Method foundm = null;
+      for(Method method : methods){
+        Class<?>[] params = method.getParameterTypes();
+        if(isCongruent(params, args)){
+          if(foundm == null){
+            foundm = method;
+            boxedArgs = boxArgs(params, args);
+          }
+        }
+      }
+      m = foundm;
+    }
+    if(m == null){
+      throw new IllegalArgumentException("No matching method found");
+    }
+    if(!Modifier.isPublic(m.getDeclaringClass().getModifiers())){
+      //public method of non-public class...
+      //Method oldm = m;
+      m = getAsMethodOfPublicBase(m.getDeclaringClass(), m);
+      if(m == null){
+        throw new IllegalArgumentException("Can't call public method of non-public class");
+      }
+    }
+    try{
+      return prepRet(m.getReturnType(), m.invoke(target, boxedArgs));
+    }catch(Exception e){
+      throw SneakyThrow.sneakyThrow(e);
+    }
+  }
+  
+  public static Object prepRet(Class<?> c, Object obj){
+    if(!(c.isPrimitive() || c == Boolean.class)){
+      return obj;
+    }
+    if(obj instanceof Boolean){
+      return ((Boolean) obj) ? Boolean.TRUE : Boolean.FALSE;
+    }
+    return obj;
+  }
+  
+  static Object boxArg(Class<?> paramType, Object arg){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.